src/deploy/functions/runtimes/node/validate.ts

```diff
@@ -0,0 +1,57 @@
+import * as path from "path";
+
+import { FirebaseError } from "../../../../error";
+import { logger } from "../../../../logger";
+import * as fsutils from "../../../../fsutils";
+
+// have to require this because no @types/cjson available
+// tslint:disable-next-line
+const cjson = require("cjson");
+
+/**
+ * Asserts that functions source directory exists and source file is present.
+ * @param data Object representing package.json file.
+ * @param sourceDir Directory for the functions source.
+ * @param projectDir Project directory.
+ * @throws { FirebaseError } Functions source directory and source file must exist.
+ */
+function assertFunctionsSourcePresent(data: any, sourceDir: string, projectDir: string): void {
+  const indexJsFile = path.join(sourceDir, data.main || "index.js");
+  if (!fsutils.fileExistsSync(indexJsFile)) {
+    const msg = `${path.relative(
+      projectDir,
+      indexJsFile
+    )} does not exist, can't deploy Cloud Functions`;
+    throw new FirebaseError(msg);
+  }
+}
+
+/**
+ * Validate contents of package.json to ensure main file is present.
+ * @param sourceDirName Name of source directory.
+ * @param sourceDir Relative path of source directory.
+ * @param projectDir Relative path of project directory.
+ * @param hasRuntimeConfigInConfig Whether the runtime was chosen in the `functions` section of firebase.json.
+ * @throws { FirebaseError } Package.json must be present and valid.
+ */
+export function packageJsonIsValid(
+  sourceDirName: string,
+  sourceDir: string,
+  projectDir: string
+): void {
+  const packageJsonFile = path.join(sourceDir, "package.json");
+  if (!fsutils.fileExistsSync(packageJsonFile)) {
+    const msg = `No npm package found in functions source directory. Please run 'npm init' inside ${sourceDirName}`;
```

Comment:
Not sure if this error message should recommend running npm init, for 2 reasons:
- I think the most common cause of this would be having an incorrect functions.source listed in firebase.json
- In general, we should push people toward using firebase init over npm init.
I'd suggest changing this to `No npm package found in functions source directory '${sourceDirName}'`